### PR TITLE
fix(type-system): fix mixed-type type inference by using pyarrow

### LINKED ISSUE #5810: bug: binary type conversion issue with memtable registration
### What happened?

Here is a different minimal example. Workaround is to convert 'UNK' into null or the entire column into string. 

```
    zc = pd.Series([ 6770, 54601, 54650, 37043, 16146,'UNK', 1550, 42211, 15601, 1440], name='zip')
    zcv = ibis.memtable(zc)
    zcv.schema()
```
comes back with: 
```
ibis.Schema {
  zip  binary
}
```

but trying to show just what's in the table `zcv` shows the below error: 
```
Traceback (most recent call last):
  File "...\envs\ml310\lib\site-packages\IPython\core\formatters.py", line 221, in catch_format_error
    r = method(self, *args, **kwargs)
  File "...\envs\ml310\lib\site-packages\IPython\core\formatters.py", line 706, in __call__
    printer.pretty(obj)
  File "...\envs\ml310\lib\site-packages\IPython\lib\pretty.py", line 410, in pretty
    return _repr_pprint(obj, self, cycle)
  File "...\envs\ml310\lib\site-packages\IPython\lib\pretty.py", line 778, in _repr_pprint
    output = repr(obj)
  File "...\envs\ml310\lib\site-packages\ibis\expr\types\core.py", line 56, in __repr__
    simple_console.print(self)
  File "...\envs\ml310\lib\site-packages\rich\console.py", line 1698, in print
    extend(render(renderable, render_options))
  File "...\envs\ml310\lib\site-packages\rich\console.py", line 1310, in render
    render_iterable = renderable.__rich_console__(self, _options)  # type: ignore[union-attr]
  File "...\envs\ml310\lib\site-packages\ibis\expr\types\relations.py", line 259, in __rich_console__
    table = to_rich_table(self, width)
  File "...\envs\ml310\lib\site-packages\ibis\expr\types\pretty.py", line 230, in to_rich_table
    result = table.limit(nrows + 1).execute()
  File "...\envs\ml310\lib\site-packages\ibis\expr\types\core.py", line 303, in execute
    return self._find_backend(use_default=True).execute(
  File "...\envs\ml310\lib\site-packages\ibis\backends\base\sql\__init__.py", line 246, in execute
    self._register_in_memory_tables(expr)
  File "...\envs\ml310\lib\site-packages\ibis\backends\base\sql\__init__.py", line 262, in _register_in_memory_tables
    self._register_in_memory_table(memtable)
  File "...\envs\ml310\lib\site-packages\ibis\backends\duckdb\__init__.py", line 695, in _register_in_memory_table
    table = op.data.to_pyarrow(schema)
  File "...\envs\ml310\lib\site-packages\ibis\backends\pandas\client.py", line 236, in to_pyarrow
    return pa.Table.from_pandas(self._df, schema=ibis_to_pyarrow_schema(schema))
  File "pyarrow\table.pxi", line 3557, in pyarrow.lib.Table.from_pandas
  File "...\envs\ml310\lib\site-packages\pyarrow\pandas_compat.py", line 611, in dataframe_to_arrays
    arrays = [convert_column(c, f)
  File "...\envs\ml310\lib\site-packages\pyarrow\pandas_compat.py", line 611, in <listcomp>
    arrays = [convert_column(c, f)
  File "...\envs\ml310\lib\site-packages\pyarrow\pandas_compat.py", line 598, in convert_column
    raise e
  File "...\envs\ml310\lib\site-packages\pyarrow\pandas_compat.py", line 592, in convert_column
    result = pa.array(col, type=type_, from_pandas=True, safe=safe)
  File "pyarrow\array.pxi", line 316, in pyarrow.lib.array
  File "pyarrow\array.pxi", line 83, in pyarrow.lib._ndarray_to_array
  File "pyarrow\error.pxi", line 123, in pyarrow.lib.check_status
pyarrow.lib.ArrowTypeError: ("Expected bytes, got a 'int' object", 'Conversion failed for column zip with type object')

```

### What version of ibis are you using?

5.0

### What backend(s) are you using, if any?

_No response_

### Relevant log output

_No response_

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

### PR DESCRIPTION
**Pull request by cpcloud**
Fixes #5810.

This PR fixes inference of pandas objects that contained mixed-type values.

Originally I implemented this without pyarrow, and things quickly become much
more complex than this PR.

@kszucs had the great idea of using pyarrow to do the inference, which has the
benefit of being both faster and results in less code.

The main potential drawback is that now pyarrow is a hard dependency. I'm
leaning towards that not being a huge issue given its prevalence and the fact
that it's becoming more of a maintenance burden to bend over backwards to avoid
it. The tradeoff seems worth it at this point.

Another issue is the snowflake backend.

Since snowflake pins pyarrow to 10, and after this we have a hard dependency on
pyarrow, it's now easier without the snowflake pandas extra to get a partially-invalid
install with `pip install ibis-framework[snowflake]`. We can prevent this by requiring
the `pandas` extra in snowflake deps, which will do the right thing with the previous pip command
while also allowing non-snowflake extras to potentially install newer versions of pyarrow.

The main downside to constraining snowflake is that developer environments are
now limited to pyarrow 10, though even that can be worked around if necessary.

### CODE REVIEW

File: pyproject.toml [hunk 1]
```diff
@@ -191,10 +193,9 @@ impala = ["fsspec", "impyla", "requests", "sqlalchemy"]
 mssql = ["sqlalchemy", "pymssql", "sqlalchemy-views"]
 mysql = ["sqlalchemy", "pymysql", "sqlalchemy-views"]
 pandas = ["regex"]
-polars = ["polars", "pyarrow"]
+polars = ["polars"]
 postgres = ["psycopg2", "sqlalchemy", "sqlalchemy-views"]
-pyarrow = ["pyarrow"]
```
Comment (cpcloud) [inline pyproject.toml:196]: I'm going to remove this before 5.1 so we don't introduce it, only to take it away in a major release.

File: ibis/backends/pyarrow/datatypes.py [hunk 2]
```diff
@@ -84,22 +84,42 @@ def from_ibis_map(dtype: dt.Map):
...
 @dt.dtype.register(pa.DataType)  # type: ignore[misc]
 def from_pyarrow_primitive(
-    arrow_type: pa.DataType,
-    nullable: bool = True,
+    arrow_type: pa.DataType, nullable: bool = True
 ) -> dt.DataType:
-    return _to_ibis_dtypes[arrow_type](nullable=nullable)
+    if (dtype := _to_ibis_dtypes.get(arrow_type)) is not None:
+        return dtype(nullable=nullable)
+    # workaround arrow 10 bug
```
Comment (kszucs) [inline ibis/backends/pyarrow/datatypes.py:113]: What is this bug?

File: ibis/expr/datatypes/value.py [hunk 3]
```diff
@@ -167,37 +167,26 @@ def __repr__(self):
...
+    import pyarrow as pa
+
+    import ibis.backends.pyarrow.datatypes  # noqa: F401
+
+    try:
+        pa_type = pa.array(x, from_pandas=True).type
+    except pa.ArrowInvalid:
+        try:
+            # handle embedded series objects
+            return dt.highest_precedence(map(infer, x))
```
Comment (kszucs) [inline ibis/expr/datatypes/value.py:179]: Do we have a test case for this?

**Review by kszucs — APPROVED** [review]
Just two questions, otherwise it looks good to me.

Nice improvement!

File: ibis/backends/pyarrow/datatypes.py (hunk 2 above)
Comment (cpcloud) [reply ibis/backends/pyarrow/datatypes.py:113]: PyArrow 11:

```
In [9]: import pyarrow as pa

In [10]: from datetime import time

In [11]: pa.array([time(0), time(1), time(2)])
Out[11]:
<pyarrow.lib.Time64Array object at 0x7fcd5635d240>
[
  00:00:00.000000,
  01:00:00.000000,
  02:00:00.000000
]

In [12]: pa.array([time(0), time(1), time(2)]).type
Out[12]: Time64Type(time64[us])

In [13]:
…/ibis on  master is 📦 v5.0.0 via 🐍 v3.10.10 via ❄️  impure (ibis-3.10.10-env) took 57s
❯ poetry show pyarrow
 name         : pyarrow
 version      : 11.0.0
 description  : Python library for Apache Arrow

dependencies
 - numpy >=1.16.6

required by
 - datafusion >=1
 - db-dtypes >=3.0.0
```

PyArrow 10:

```
In [1]: import pyarrow as pa

In [2]: from datetime import time

In [3]: pa.array([time(0), time(1), time(2)]).type
Out[3]: DataType(time64[us])

…/ibis on  fix-memtable-inference is 📦 v5.0.0 via 🐍 v3.10.10 via ❄️  impure (ibis-3.10.10-env) took 18s
❯ poetry show pyarrow
 name         : pyarrow
 version      : 10.0.1
 description  : Python library for Apache Arrow

dependencies
 - numpy >=1.16.6

required by
 - datafusion >=1
 - db-dtypes >=3.0.0
 - snowflake-connector-python >=10.0.1,<10.1.0
```

**Pull request merged**